File: src/bin/day-1.py
import pathlib

file_path = pathlib.Path("src/bin/day-1-input.txt")

ELVES_CALORIES = [0]

with open(file_path, "r") as f:
    for line in f:
        if line.strip():
            ELVES_CALORIES[-1] += int(line.strip())
        else:
            ELVES_CALORIES.append(0)


ELF_WITH_MOST_CALORIES = 0
for i, calories in enumerate(ELVES_CALORIES):
    if calories > ELVES_CALORIES[ELF_WITH_MOST_CALORIES]:
        ELF_WITH_MOST_CALORIES = i
print(f"Elf with most calories: {ELF_WITH_MOST_CALORIES + 1} ({max(ELVES_CALORIES)} calories)")

TOP_3_CALORIES = sum(sorted(ELVES_CALORIES, reverse=True)[:3])
print(f"Top 3 elves: {TOP_3_CALORIES} calories")

File: src/bin/day-1.rs
use std::cmp::Ordering;
use std::fs;

fn get_total_calories(path: &str) -> Vec<u32> {
    let contents = fs::read_to_string(path).expect("Something went wrong reading the file");
    let mut total_calories: Vec<u32> = Vec::new();
    let mut calories = 0;
    for line in contents.lines() {
        match line.trim().parse::<u32>() {
            Ok(num) => calories += num,
            Err(_) => {
                total_calories.push(calories);
                calories = 0;
            }
        };
    }
    total_calories
}

fn find_max_calories(total_calories: &Vec<u32>) -> u32 {
    let mut max_calories = 0;
    for calories in total_calories {
        match calories.cmp(&max_calories) {
            Ordering::Greater => max_calories = *calories,
            _ => (),
        }
    }
    max_calories
}

fn main() {
    let path = "src/bin/day-1-input.txt";
    let mut total_calories = get_total_calories(path);
    println!("Elf with most calories: {}", find_max_calories(&total_calories));
    total_calories.sort();
    total_calories.reverse();
    println!("Elf with most calories: {}", total_calories[0]);
    println!("Top 3 elves: {}", total_calories[..3].iter().sum::<u32>());
}
